MWAppAuthPlugin/MWAppAuthPlugin/AppAuthStep/MWAppAuthStepViewController.swift

```diff
@@ -161,8 +166,165 @@ extension MWAppAuthStepViewController: MobileWorkflowButtonTableViewCellDelegate
+            DispatchQueue.main.async {
+                self?.hideLoading()
+                switch response {
+                case .success:
+                    self?.goForward()
+                case .failure(let error):
+                    self?.show(error)
+                }
+            }
+        }
+    }
+    
+    func makeName(fullName: PersonNameComponents?) -> String? {
+        guard let fullName = fullName else {
+            return nil
+        }
+        
+        var nameComponents = [String]()
+        if let givenName = fullName.givenName {
+            nameComponents.append(givenName)
+        }
+        if let familyName = fullName.familyName {
+            nameComponents.append(familyName)
+        }
+        
+        return nameComponents.joined(separator: " ")
+    }
+}
+
+struct AppleIDCredential: Codable {
+    let userId: String
+    let name: String
+    let identityToken: String
+    
+    enum CodingKeys: String, CodingKey {
+        case userId = "user_identity"
+        case name = "name"
+        case identityToken = "jwt"
+    }
+}
+
+// MARK: - Authorization Controller Delegate
+
+extension MWAppAuthStepViewController: ASAuthorizationControllerDelegate {
+    
+    func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
+        
+        switch authorization.credential {
+        case let appleIDCredential as ASAuthorizationAppleIDCredential:
+            
+            // Expiration date is not currently used, hence credential was set to `.distantFuture`. This should be reviewed in the future.
+            let userIdCredential = Credential(
+                type: CredentialType.appleIdCredentialUser.rawValue,
+                value: appleIDCredential.user,
+                expirationDate: .distantFuture
+            )
+
+            self.appAuthStep.services.credentialStore.updateCredential(userIdCredential) { [weak self] result in
+                switch result {
+                case .success:
+                    guard let identityToken = appleIDCredential.identityToken, let identityTokenString = String(data: identityToken, encoding: .utf8) else {
+                        return
+                    }
+                    
+                    let name = self?.makeName(fullName: appleIDCredential.fullName) ?? ""
+                    
+                    self?.performSignInWithApple(userId: appleIDCredential.user, name: name, identityToken: identityTokenString)
+                    
+                case .failure(let error):
+                    self?.show(error)
+                }
+            }
+            
+        default:
+            break
         }
     }
+    
+    func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Error) {
+        guard let authError = error as? ASAuthorizationError, authError.code != .canceled else {
+            return
```

Comment:
Again, `preconditionFailure` better here if we don't expect it to happen?

Reply:
In this case I would leave the return, since if I click on Login with Apple, and then cancel button, the `authError.code` would be `.canceled`. I just `guard let` there to show an error (except when user cancelled the sign in voluntarily).

Reply:
Ah yes sorry, didn't parse it fully and missed the cancel check.